### PR TITLE
[Tests] Nullable

### PR DESCRIPTION
**Pull request by ysolomchenko**
Towards https://github.com/open-telemetry/opentelemetry-dotnet/issues/3958

## Changes

Tests mark as nullable

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: test/OpenTelemetry.Tests/Logs/BatchLogRecordExportProcessorTests.cs [hunk 1]
```diff
@@ -43,14 +43,14 @@ public void StateValuesAndScopeBufferingTest()
...
         Assert.Same("Value", actualState.Key);
         Assert.Same("Hello world", actualState.Value);
 
         bool foundScope = false;
 
-        logRecord.ForEachScope<object>(
+        logRecord.ForEachScope<object?>(
             (s, o) =>
             {
                 foundScope = ReferenceEquals(s.Scope, exportedItems);
```
Comment (TimothyMothra) [inline test/OpenTelemetry.Tests/Logs/BatchLogRecordExportProcessorTests.cs:56]: Unrelated to your changes, but this looks like a test bug.

The foreach method is setting a boolean. This means only the last scope inspected is recorded.
I think this instead should be using an on-equals `|=`

```suggestion
                foundScope |= ReferenceEquals(s.Scope, exportedItems);
```

File: test/OpenTelemetry.Tests/Metrics/MetricPointReclaimTestsBase.cs [hunk 2]
```diff
@@ -28,7 +28,7 @@ protected MetricPointReclaimTestsBase(bool emitOverflowAttribute)
         }
 
         this.configuration = new ConfigurationBuilder()
-            .AddInMemoryCollection(this.configurationData)
+            .AddInMemoryCollection(this.configurationData!)
```
Comment (TimothyMothra) [inline test/OpenTelemetry.Tests/Metrics/MetricPointReclaimTestsBase.cs:31]: configurationData is not nullable. Why is this change needed?

File: test/OpenTelemetry.Tests/TestSelfDiagnosticsConfigRefresher.cs [hunk 3]
```diff
@@ -6,16 +6,16 @@
...
     public bool TryGetLogStreamCalled { get; private set; }
 
     public override bool TryGetLogStream(int byteCount, [NotNullWhen(true)] out Stream stream, out int availableByteCount)
     {
         this.TryGetLogStreamCalled = true;
-        stream = this.stream;
+        stream = this.stream!;
         availableByteCount = 0;
         return true;
     }
```
Comment (TimothyMothra) [inline test/OpenTelemetry.Tests/TestSelfDiagnosticsConfigRefresher.cs:21]: This method states `stream` is not null when the method returns true.
But I don't see that check ever being performed. 

- I think instead the out param needs to be `Stream?`
- remove the null-forgiving operator `!` on line 18
- change line 20 to `return this.stream != null`

Comment (ysolomchenko) [reply test/OpenTelemetry.Tests/TestSelfDiagnosticsConfigRefresher.cs]: fixed in b0cbbe775bfa80d9a901c168ecdfeebbb1db0dd0

File: test/OpenTelemetry.Tests/Metrics/MetricPointReclaimTestsBase.cs (hunk 2 above)
Comment (ysolomchenko) [reply test/OpenTelemetry.Tests/Metrics/MetricPointReclaimTestsBase.cs]: fixed in b0cbbe775bfa80d9a901c168ecdfeebbb1db0dd0

**Review by TimothyMothra — APPROVED** [review]
LGTM

**Review by Kielek — APPROVED** [review]
